fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/PartitionColPredicateExtractor.java

```diff
@@ -0,0 +1,193 @@
+import com.starrocks.sql.optimizer.operator.scalar.LikePredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator.BinaryType.EQ_FOR_NULL;
+
+/**
+ * Extract predicate tree only with partition key or boolean constant
+ * from original predicate tree.
+ * For example, k1 is partition key and c1, c2 are normal column.
+ * This predicate (k1 < 100 and c1 like 'a') or (k1 > 200 and func(k1) > c1)
+ * will transform to
+ * k1 < 100 or k1 > 200
+ * Then we can use range info to test which ranges fit the transformed condition.
+ */
+public class PartitionColPredicateExtractor extends ScalarOperatorVisitor<ScalarOperator, Void> {
+
+    private ColumnRefSet partitionColumnSet;
+
+    public PartitionColPredicateExtractor(RangePartitionInfo rangePartitionInfo,
+                                          Map<Column, ColumnRefOperator> columnMetaToColRefMap) {
+        List<ColumnRefOperator> columnRefOperators = Lists.newArrayList();
+        Column partitionColumn = rangePartitionInfo.getPartitionColumns().get(0);
+        columnRefOperators.add(columnMetaToColRefMap.get(partitionColumn));
+        partitionColumnSet = new ColumnRefSet(columnRefOperators);
+    }
+
+    public ScalarOperator extract(ScalarOperator predicate) {
+        predicate = predicate.clone();
+        List<ScalarOperator> splitPredicates = Utils.extractConjuncts(predicate).stream()
+                .filter(e -> !e.isFromPredicateRangeDerive()).collect(Collectors.toList());
+        ScalarOperator scalarOperator = Utils.compoundAnd(splitPredicates);
+        return scalarOperator.accept(this, null);
+    }
+
+    @Override
+    public ScalarOperator visit(ScalarOperator scalarOperator, Void context) {
+        return null;
+    }
+
+    @Override
+    public ScalarOperator visitConstant(ConstantOperator literal, Void context) {
+        return literal;
+    }
+
+    // only column is partition column will be preserved
+    @Override
+    public ScalarOperator visitVariableReference(ColumnRefOperator variable, Void context) {
+        if (partitionColumnSet.containsAll(variable.getUsedColumns())) {
+            return variable;
+        }
+        return null;
+    }
+
+    @Override
+    public ScalarOperator visitCall(CallOperator call, Void context) {
+        return null;
+    }
+
+    @Override
+    public ScalarOperator visitBinaryPredicate(BinaryPredicateOperator predicate, Void context) {
+        if (partitionColumnSet.containsAll(predicate.getUsedColumns())) {
+            ScalarOperator left = predicate.getChild(0).accept(this, null);
+            if (isShortCut(left)) {
+                return ConstantOperator.createBoolean(true);
```

Comment:
If return true constant, which means we need to keep all partitions or prunes all partitions?